### ISSUE TITLE
`-iframework` wrongly passed to certain targets of GCC on Darwin host

### ISSUE THREAD

**Issue by Ivan-Velickovic** [NixOS/nixpkgs#257258]
### Describe the bug

When building certain ARM GCC cross-compilers on an x86 Darwin host, the `-iframework` flag is passed erroneously. The two cross-compilers that I have reproduced the issue for is `aarch64-none-elf` and `aarch64-unknown-linux-gnu`. Neither of these is targeting Darwin/macOS and hence the `-iframework` flag is expected to be unrecognisable to GCC.

I do not have access to an ARM64 Darwin machine, so do not know if the issue is reproducible there.

### Steps To Reproduce

1. Enter a Nix shell using the cross-compiler with `nix-shell -p pkgsCross.aarch64-embedded.buildPackages.gcc` or `nix-shell -p pkgsCross.aarch64-multiplatform.buildPackages.gcc`
2. Run any command invoking the compiler, such as `aarch64-none-elf-gcc --version`
3. Observe the following behaviour: `aarch64-none-elf-gcc: error: unrecognized command-line option ‘-iframework’`

### Expected behavior
No error message about `-iframework`.

### Notify maintainers

@vcunat @Ericson2314

### Metadata
```console
 - system: `"x86_64-darwin"`
 - host os: `Darwin 21.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.16.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

**Comment by Ivan-Velickovic** [comment]
Fairly new to Nix, apologies if I've gotten anything wrong.

**Comment by Ivan-Velickovic** [comment]
Pinging @Ericson2314.

**Comment by Ivan-Velickovic** [comment]
Also observed this when trying to get a RISC-V toolchain as well.

**Comment by wucke13** [comment]
This is still relevant. Similar issues (some same cause, some with different cause):

- https://github.com/NixOS/nixpkgs/issues/339245
- https://github.com/NixOS/nixpkgs/issues/283466 (User spliced in the `pkgsCross` by hand, so user error but same symptom)
- https://github.com/NixOS/nixpkgs/issues/248206 (same as above)
- https://github.com/NixOS/nixpkgs/issues/48205 (same as above)

**Comment by Liamolucko** [comment]
This seems to have been fixed by #346043, since there's no longer anything in `stdenv` which contains `$out/Library/Frameworks` and triggers this: https://github.com/NixOS/nixpkgs/blob/b0fb9600be0838cbb8fcd7679527c65ff1070cdc/pkgs/build-support/cc-wrapper/setup-hook.sh#L75-L77

Are there any other `.framework`s in nixpkgs which still need that hook, or can it be removed now?

As for the more fundamental reason for the bug, I think what's happening is that when you put a cross compiler into `nativeBuildInputs`, you're saying to `stdenv` 'I declare that this package runs on this derivation's build platform, and targets this derivation's host platform'; and so as a result, it gives the cross compiler all the things which run on the host platform (`buildInputs`, which I think used to include `darwin.CF` by default) to link against, include headers from, etc. As a workaround, you can put the compiler in `depsBuildTarget` instead, since you probably don't have anything in `depsTargetTarget` for `stdenv` to try and pass to the cross compiler.

**Comment by Ivan-Velickovic** [comment]
Yes, that PR seems to have fixed. Thanks!
